README.md
# wave_transition

The package will provide smooth wave transitions between pages without having to write any boilerplate code.

![example](https://user-images.githubusercontent.com/67534990/108588638-48d72700-7380-11eb-958f-371f8b7d7b27.gif)

## Getting Started

[Example] (https://github.com/sonykurian96/wave_transition/blob/master/example/lib/main.dart)

To use this package : *add the dependency to your [pubspec.yaml] file

```yaml
   dependencies:
     flutter:
       sdk: flutter
     wave_transition:
```

## use case

```dart
   onPressed: () {
     Navigator.push(
       context,
       WaveTransition(
        child: SecondScreen(),
        center: FractionalOffset(0.90, 0.90)
        duration: Duration(milliseconds: 3000) // optional
      )
   );
 },
```

You can pass arguments using RouteSettings ->

```dart
    settings: RouteSettings(
      arguments: "yeah! it works!"
    )
```


CHANGELOG.md
## [0.0.5] - 24/07/2021

* Updated license file.

## [0.0.4] - 12/06/2021

* Supports null safety.

## [0.0.3] - 01/03/2021

* Added functionality to pass arguments.

## [0.0.2] - Modified README.md

* Example gif file Added.

## [0.0.1] - Initial release

* Starting point.

lib/transition.dart
import 'package:flutter/material.dart';

class Transition extends StatelessWidget {

  // secondary widget
  Transition(
      {required this.targetPage, required this.fromCenter, this.duration});

  // target widget    
  final Widget targetPage;

  // center of transition
  final FractionalOffset fromCenter;

  // duration of transition
  final Duration? duration;

  @override
  Widget build(BuildContext context) {
    return TweenAnimationBuilder(
      tween: Tween(begin: 0.0, end: 1.0),
      duration: duration ?? Duration(milliseconds: 1500),
      child: targetPage,
      builder: (context, dynamic value, child) {
        return ShaderMask(
          blendMode: BlendMode.modulate,
          shaderCallback: (rect) {
            return RadialGradient(
              colors: [
                Colors.white,
                Colors.white,
                Colors.transparent,
                Colors.transparent
              ],
              stops: [0.0, 0.55, 0.6, 1.0],
              center: fromCenter,
              radius: value * 5,
            ).createShader(rect);
          },
          child: child,
        );
      },
    );
  }
}


lib/wave_transition.dart
library wave_transition;

/// This package allows you to create amazing wave transitions///

import 'package:flutter/material.dart';
import 'transition.dart';

class WaveTransition<T> extends PageRouteBuilder<T> {

  // target widget should be passed here
  final Widget child;

  // set alignment 
  final FractionalOffset center;

  // duration of animation
  final Duration? duration;

  // main widget
  WaveTransition(
      {required this.child,
      required this.center,
      this.duration,
      RouteSettings? settings})
      : super(
          settings: settings,
          opaque: false,
          pageBuilder: (context, animation, _) {
            return Transition(
              targetPage: child,
              fromCenter: center,
              duration: duration,
            );
          },
        );
}


example/lib/customWidget.dart
import 'package:flutter/material.dart';

// ignore: must_be_immutable
class Custom extends StatelessWidget {
  Custom(
      {required this.answer,
      required this.videoLink,
      required this.question,
      required this.customRaisedButton,
      required this.isVisible});

  String videoLink;
  String question;
  String answer;
  bool isVisible;
  Widget customRaisedButton;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(10, 0, 10, 0),
      child: Card(
        elevation: 20,
        shadowColor: Colors.amber,
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(30),
        ),
        child: Card(
          shadowColor: Colors.amber,
          child: Column(
            children: [
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Text(
                  question,
                  style: TextStyle(fontSize: 20),
                ),
              ),
              Divider(thickness: 1),
              ListTile(
                leading: RaisedButton.icon(
                  icon: Icon(Icons.ondemand_video, color: Colors.orange[500]),
                  textColor: Colors.orange[500],
                  color: Colors.white,
                  label: Text("Watch Video"),
                  onPressed: () {},
                  shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(30.0),
                      side: BorderSide(color: Colors.orange[500]!)),
                ),
                trailing: customRaisedButton,
              ),
              Divider(thickness: 1),
              Visibility(
                visible: isVisible,
                child: Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Text(
                    answer,
                    style: TextStyle(fontSize: 16),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


example/lib/secondScreen.dart
import 'package:flutter/material.dart';
import 'customWidget.dart';

class SecondScreen extends StatefulWidget {
  @override
  _SecondScreenState createState() => _SecondScreenState();
}

class _SecondScreenState extends State<SecondScreen> {
  Widget button() {
    return RaisedButton(
      textColor: Colors.white,
      color: Colors.orange[500],
      child: Text("View Answer"),
      onPressed: () {},
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(30.0),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    String? _data = ModalRoute.of(context)!.settings.arguments as String?;
    print(_data);
    return Scaffold(
      appBar: AppBar(
        title: Text("Enjoy!!!"),
        centerTitle: true,
        backgroundColor: Colors.orange[500],
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            Custom(
              videoLink: "https://youtu.be/bxRVz8zklWM",
              isVisible: true,
              //displayAnswer,
              customRaisedButton: button(),
              question:
                  "Define and explain the stack data structure with suitable example",
              answer: """Answer:- 
                        A stack is an Abstract Data Type (ADT), commonly used in most programming languages. It is named stack as it behaves like a real-world stack, for example – a deck of cards or a pile of plates, etc.
Stack Example
A real-world stack allows operations at one end only. For example, we can place or remove a card or plate from the top of the stack only. Likewise, Stack ADT allows all data operations at one end only. At any given time, we can only access the top element of a stack.""",
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'secondScreen.dart';
import 'package:wave_transition/wave_transition.dart';

void main() {
  runApp(
    MaterialApp(
      debugShowCheckedModeBanner: false,
      home: FirstScreen(),
    ),
  );
}

class FirstScreen extends StatefulWidget {
  @override
  _FirstScreenState createState() => _FirstScreenState();
}

class _FirstScreenState extends State<FirstScreen> {
  bool isSwitched = false;

  bool? isChecked1 = false;
  bool? isChecked2 = false;
  bool isAscending = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Home Page"),
        backgroundColor: Colors.orange[500],
      ),
      backgroundColor: Colors.white,
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.fromLTRB(20, 2, 0, 0),
            child: Text(
              "Last Moment Prep",
              style: TextStyle(
                fontSize: 20,
              ),
            ),
          ),
          Divider(thickness: 1),
          ListTile(
            leading: IconButton(
                icon: Icon(
                  Icons.adjust,
                  color: Colors.orange[500],
                ),
                onPressed: () {
                  print("object");
                }),
            title: Text("Sort Questions"),
            trailing: Switch(
              value: isSwitched,
              onChanged: (value) {
                setState(() {
                  isSwitched = value;
                  print(isSwitched);
                });
              },
              activeColor: Colors.green,
            ),
          ),
          Visibility(
            visible: true, //isSwitched,
            child: RaisedButton(
              textColor: Colors.white,
              color: Colors.orange[500],
              child: Text("Howdy"),
              onPressed: () {},
              shape: new RoundedRectangleBorder(
                borderRadius: new BorderRadius.circular(30.0),
              ),
            ),
          ),
          Visibility(
            visible: true, //isSwitched,
            child: Card(
              shadowColor: Colors.amber,
              elevation: 10,
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(20.0),
              ),
              child: Column(
                children: [
                  Padding(
                    padding: const EdgeInsets.fromLTRB(0, 5, 0, 5),
                    child: Text(
                      "Select type of sorting: ",
                      style: TextStyle(
                        fontSize: 20,
                      ),
                    ),
                  ),
                  Divider(
                    thickness: 1,
                  ),
                  CheckboxListTile(
                    title: Text("According to difficulty level"),
                    controlAffinity: ListTileControlAffinity.trailing,
                    onChanged: (bool? value) {
                      setState(() {
                        isChecked1 = value;
                      });
                    },
                    value: isChecked1,
                  ),
                  CheckboxListTile(
                    title: Text("According to weightage"),
                    controlAffinity: ListTileControlAffinity.trailing,
                    onChanged: (bool? value) {
                      setState(() {
                        isChecked2 = value;
                      });
                    },
                    value: isChecked2,
                  ),
                  Divider(
                    thickness: 1,
                  ),
                  ListTile(
                    title: Text("Ascending order"),
                    trailing: Switch(
                      value: isAscending,
                      onChanged: (value) {
                        setState(() {
                          isAscending = value;
                        });
                      },
                      activeColor: Colors.green,
                    ),
                  )
                ],
              ),
            ),
          )
        ],
      ),
      floatingActionButton: FloatingActionButton(
        child: Text("Go"),
        backgroundColor: Colors.orange[500],
        onPressed: () {
          Navigator.push(
            context,
            WaveTransition(
                child: SecondScreen(),
                center: FractionalOffset(0.90, 0.90),
                duration: Duration(milliseconds: 3000),
                settings: RouteSettings(arguments: "yeah! it works!")),
          );
        },
      ),
    );
  }
}
